### DISCUSSION TITLE
alternative.platforms - Qubes

### DISCUSSION THREAD

**Discussion by ros-mac** [raspiblitz/raspiblitz#3909] [Show and tell]
Might be useful to add this to 'alternative.platforms'. Having used RaspberryPIs, VirtualBox and Proxmox, I find Qubes very convenient for running nodes with external HDDs. Quick clone of Debian base template every time you upgrade and you can revert to snapshots created at each shutdown! 

Off-topic, but in the spirit of bitcoin education, I installed Umbrel via this same process with "-s -- --install-path /mnt/data/umbrel" option (old umbrel folder pre-moved to umbrel.bak and 'umbrel/.env' + db/tor/app-data folders re-added after install finished and Umbrel-startup+docker services stopped). 

This was thrown together quickly and surely has errors;

Qubes4.1.1
If applicable, in MENU of any currently installed RaspiBlitz go to UPDATE->RELEASE to prepare for an update or reinstall.
Create StandaloneVM in PVH mode from Debian11 template and sys-firewall. Set RAM (800MB/8GB in my case); 20GB+ Private Storage; 20GB+ System Storage(allows to add an extra swapfile on fast internal SSD) in the Qube's Settings. Set storage settings high if you want to run/try all features as with LVM Thin you are not using unused space anyway (this assumes the Blockchain is on an external drive). My 2/3 of features install had 14GB used on Private Storage and 6GB+5GBswapfile on System Storage.

Login, download script, run script;
    wget https://raw.githubusercontent.com/rootzoll/raspiblitz/v1.8/build_sdcard.sh   # dev script did not work for me a few days ago
    sudo bash build_sdcard.sh -f true -b v1.8 -d headless -t false -w off
Reboot qube. Attach external drive.

If applicable, create ext4 filesystem and label it to match the node name for convenience if running a few nodes. In my case I used an external with blockchain on it that I migrated from a PI to VirtualBox beforehand. 
   mkdir  /mnt/hdd 
   lsblk -f   # copy UUID for external partition
   sudo vi /etc/fstab   # add UUID=xxxxxxxxxxxxxxxxx /mnt/hdd ext4 defaults 0 1
   mount /mnt/hdd
   sudo su - admin  # this should activate the install/reinstall/update. Open more terminal tabs and do same as needed.
   tail -f raspiblitz.log   # from another tab logged in as admin

On reboot, the qube will shutdown. Make sure the external drive is pre-attached correctly before starting it again with the following commands in Dom0;
    qvm-block   # list drives and pre-attachments. qvm-usb alternative command in certain scenarios using USB major/minor numbers
    qvm-block d <qube-name> sys-usb:sdX  # removes any old persistence that may have switched block device before re-attaching
    qvm-block a  --persistent <qube-name> sys-usb:sdX   # pre-attachs external drive and will be remembered if not unattached 

Start qube again and it should automount the external drive partition from /etc/fstab based on UUID as block device might change.

Biggest negative is block devices changing in Dom0 and messing up the pre-attachments over time if sys-usb restarts, drives reconnect or Qubes itself is restarted. This can likely be mitigated with udev scripts that create another block device name by matching the external drive model details. I did this in VirtualBox successfully, but in Qubes it depends on what type of sys-usb template you are using to allow for persistence in /etc/udev/rules.d/<mapping-filename>. My sys-usb is a disposable template on top of a Debian template and I have to look into it a bit more. These worked for me in VirtualBox as a starting point.

KERNEL=="sd*", SUBSYSTEMS=="scsi", ATTRS{model}=="My Drive 686869", SYMLINK+="sdu1%n"
KERNEL=="sd*", SUBSYSTEMS=="scsi", ATTRS{model}=="My Drive X7667", SYMLINK+="sdu2%n"
KERNEL=="sd*", SUBSYSTEMS=="scsi", ATTRS{model}=="Extreme SSD", SYMLINK+="sdr1%n"
KERNEL=="sd*", SUBSYSTEMS=="scsi", ATTRS{model}=="Crucial SSD X10", SYMLINK+="sdr2%n"

Snapshots:    https://www.qubes-os.org/doc/volume-backup-revert/
I set my root and private storage for each node to 3 snapshots that I can revert to.

**Comment by BitromNO** [comment]
Ive been diving into proxmox the last days and really like what i see so far, now i have to test qubes also! thanks for the guide! :)
